### ISSUE TITLE
Axes.axline: allow to specify position in axes coordinates instead of data coordinates

### ISSUE THREAD

**Issue by johan12345** [matplotlib/matplotlib#18625]
### Problem

The `axline` function allows to plot an infinite diagonal line by specifying two points or a point and a slope in data coordinates.
There are use cases for this function where one might rather want to specify a point in axes coordinates, but keep the slope in data coordinates - e.g. when you would like to create a grid of diagonal lines filling the current viewport to highlight a diagonal feature in the plotted data.

It would probably be intuitive to allow to pass a `transform` argument to the `axline` function to accomplish this, but this raises a `ValueError: 'transform' is not allowed as a kwarg; axvline generates its own transform.`

### Proposed Solution

One could solve this either with a separate function for this use case or simply allow to pass the `transform` parameter to the `axline` function and use the provided transformation for the point, but not for the slope.

### Additional context and prior art

I have implemented a proof-of-concept solution here, based on the implementation of `axline`:

```python
def plot_diagonal(ax: Axes, xy1: tuple, slope: float, **kwargs):
    """
    Plots a diagonal line specified by one point and a slope.

    This function is similar to :meth:`~matplotlib.axes.Axes.axline`, but with the difference that the point is
    specified in axes coordinates instead of data coordinates (while the slope is still in data coordinates). That means
    that the lines stay at the same position when panning the axis.

    :param ax: Axes
    :param xy1: position of one point on the line, in axes coordinates
    :param slope: slope of the line, in data coordinates
    :param kwargs: Additional arguments to pass to the :class:`~matplotlib.lines.Line2D` constructor
    :return: :class:`~matplotlib.lines.Line2D` instance
    """
    class DiagonalLine(Line2D):
        def __init__(self, xy1, slope, **kwargs):
            super().__init__([0, 1], [0, 1], **kwargs)
            self._slope = slope
            self._xy1 = xy1

        def get_transform(self):
            ax = self.axes
            (vxlo, vylo), (vxhi, vyhi) = ax.transScale.transform(ax.viewLim)
            (x, y), = (ax.transAxes + ax.transData.inverted()).transform([self._xy1])

            # General case: find intersections with view limits in either
            # direction, and draw between the middle two points.
            _, start, stop, _ = sorted([
                (vxlo, y + (vxlo - x) * slope),
                (vxhi, y + (vxhi - x) * slope),
                (x + (vylo - y) / slope, vylo),
                (x + (vyhi - y) / slope, vyhi),
            ])
            return BboxTransformTo(Bbox([start, stop])) + ax.transLimits + ax.transAxes

        def draw(self, renderer):
            self._transformed_path = None  # Force regen.
            super().draw(renderer)

    line = DiagonalLine(xy1, slope, **kwargs)
    ax.add_line(line)
    return line

# example:
fig, ax = plt.subplots()
slope = 0.5
for pos in np.linspace(-2, 1, 10):
    # plot diagonal lines
    plot_diagonal(ax, (pos, 0), slope, color='k')
plt.show()
```

![Figure_5](https://user-images.githubusercontent.com/5310424/94815615-07f36b80-03fb-11eb-9a1f-29b42c51d905.png)

**Comment by dopplershift** [comment]
For the skew axes case, the slope is done in pixel coords, not data coords, so a bit different.

**Comment by johan12345** [comment]
The solution I posted above does work when resizing the figure. I based it upon the implementation used in `axline`, where the `draw` method is overridden to clear the cached transformed path.

In this case I use `ax.transAxes + ax.transData.inverted()` to transform the point from axis into data coordinates and then continue with the same algorithm that `axline` uses. In principle that could also be done with any other transform passed by the user, right?

**Comment by tacaswell** [comment]
Ah, I missed the `self._transformed_path = None  # Force regen.` line :sheep: 

Comparing the code in the OP to the code in `_AxLine` it is not immediately clear to me how to parameterize that code to drop in a user supplied transform in a way that  is document able (given that the input is already pretty flexible to account for "point + slope" vs "two points").

**Comment by johan12345** [comment]
Hm, I tried to implement it and it is indeed complicated to make it part of `axline`, not only because of the flexibility of "point + slope" vs. "two points", but also because when I pass `ax.transAxes` to the function it is not updated when the axes limits change. Instead, I built a function for this specific case ("point + slope", point in axes coordinates, slope in data coordinates) in #18647. Not sure about a clear naming of the function though...

**Comment by timhoffm** [comment]
We've discussed this on the dev call:

This should become part of the `axline()` API, using the `transform` parameter. `transform` should apply to the points, but not to the slope. For now, slope (if given) will always stay in data space (one could add a slope_transform later if needed).

Your use case will then become `ax.axline(xy, slope=slope, transform=ax.transAxes)`.

**Comment by johan12345** [comment]
@timhoffm Thanks!

Okay, I tried adapting the `axline` API to support this additional feature instead of making it a separate function, and it seems to work now (see updated PR #18647). What I'm not sure about is [the adjustment of data limits](https://github.com/matplotlib/matplotlib/pull/18647/files#diff-608e2a1bf3d32fd631b0763599a3334d59822c11cb4fec1fa7398c9714592900R860). For the moment, I set it to simply not adjust the data limits if a transform is passed (because at least for this use case I think it doesn't really make sense to adjust them).

Does that look good to you?
